File: tracify_backend/backend/app/schemas/credit.py
"""
Credit system schemas for Tracerfy Backend
"""

from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field
from decimal import Decimal
from enum import Enum


class CreditTransactionType(str, Enum):
    PURCHASE = "purchase"
    USAGE = "usage"
    REFUND = "refund"
    BONUS = "bonus"


class CreditPackage(BaseModel):
    id: int
    name: str
    credits: int
    price: Decimal
    description: Optional[str] = None
    bonus_credits: int = 0
    is_active: bool = True
    
    class Config:
        from_attributes = True


class CreditBalance(BaseModel):
    user_id: int
    current_credits: int
    total_purchased: int
    total_used: int
    total_bonus: int
    effective_rate: Decimal
    last_updated: datetime
    
    class Config:
        from_attributes = True


class CreditTransaction(BaseModel):
    id: int
    user_id: int
    transaction_type: CreditTransactionType
    amount: int
    balance_after: int
    description: Optional[str] = None
    reference_id: Optional[str] = None
    created_at: datetime
    
    class Config:
        from_attributes = True


class CreditPurchase(BaseModel):
    package_id: int
    payment_method_id: str
    amount: Decimal
    currency: str = "USD"


class CreditPurchaseResponse(BaseModel):
    transaction_id: str
    credits_added: int
    bonus_credits: int
    total_credits: int
    new_balance: int
    amount_charged: Decimal
    currency: str


class CreditUsage(BaseModel):
    amount: int
    description: Optional[str] = None
    reference_id: Optional[str] = None


class CreditUsageResponse(BaseModel):
    transaction_id: str
    credits_used: int
    new_balance: int
    description: str


class CreditTransactionHistory(BaseModel):
    transactions: List[CreditTransaction]
    total_count: int
    page: int
    per_page: int


class CreditStats(BaseModel):
    current_balance: int
    total_purchased: int
    total_used: int
    total_bonus: int
    effective_rate_per_credit: Decimal
    last_purchase_date: Optional[datetime] = None
    last_usage_date: Optional[datetime] = None
    usage_this_month: int
    purchases_this_month: int

File: exact-site-replica-main/src/types/credit.ts
export interface CreditPackage {
  id: number;
  name: string;
  credits: number;
  price: number;
  description?: string;
  bonus_credits: number;
  is_active: boolean;
}

export interface CreditBalance {
  user_id: number;
  current_credits: number;
  total_purchased: number;
  total_used: number;
  total_bonus: number;
  effective_rate: number;
  last_updated: string;
}

export interface CreditTransaction {
  id: number;
  user_id: number;
  transaction_type: 'purchase' | 'usage' | 'refund' | 'bonus';
  amount: number;
  balance_after: number;
  description?: string;
  reference_id?: string;
  created_at: string;
}

export interface CreditPurchase {
  package_id: number;
  payment_method_id: string;
  amount: number;
  currency: string;
}

export interface CreditPurchaseResponse {
  transaction_id: string;
  credits_added: number;
  bonus_credits: number;
  total_credits: number;
  new_balance: number;
  amount_charged: number;
  currency: string;
}

export interface CreditUsage {
  amount: number;
  description?: string;
  reference_id?: string;
}

export interface CreditUsageResponse {
  transaction_id: string;
  credits_used: number;
  new_balance: number;
  description: string;
}

export interface CreditTransactionHistory {
  transactions: CreditTransaction[];
  total_count: number;
  page: number;
  per_page: number;
}

export interface CreditStats {
  current_balance: number;
  total_purchased: number;
  total_used: number;
  total_bonus: number;
  effective_rate_per_credit: number;
  last_purchase_date?: string;
  last_usage_date?: string;
  usage_this_month: number;
  purchases_this_month: number;
}
